### ISSUE TITLE
corporate github repo: not generating events and unable to write commit status to github

### ISSUE THREAD

**Issue by sairamvoggu1007** [fluxcd/flux2#1106]
Team,

events are not generating and unable to write commit status to github about the deployment

Following is error noticing in notification controller log.. It was supposed to fetch from my github domain but it's trying to fetch from api.github.com please check following error
 **```
https://<mygithubdomain>/s-voggu/eks but it's trying to GET from Get "https://api.github.com/repos/s-voggu/eks/. please check below
{"level":"error","ts":"2021-03-12T18:40:29.577Z","logger":"event-server","msg":"failed to send notification","object":"platform/badlybehavingtestservice","kind":"Kustomizatio │
│ n","error":"could not list commit statuses: Get \"https://api.github.com/repos/s-voggu/eks/commits/d1cc91e17c740a4d203e8eef04b8312bab47f46b/statuses?per_page=50\": net/http:
```**


In gitrepository type resource I'm using ssh repo url and able to fetch latest commit status...In providers I'm using https repo url in address field.  In gitrespositry resource and kusomization resource events are getting generated but I don't see any events in alerts and providers . Observing above error in notification controller log please assist asap

**Comment by simongottschlag** [comment]
Hi!

There seems to be some missing information in the log you posted after net/http.

Can you confirm the you are able to access the api from a pod and nothing is blocking it?

**Comment by sairamvoggu1007** [comment]
`{"level":"error","ts":"2021-03-16T02:22:48.760Z","logger":"event-server","msg":"failed to send notification","object":"platform/orderstatusfeedservice","kind":"Kustomization","error":"could not list commit statuses: Get \"https://api.github.com/repos/s-voggu/eks/commits/17e02455855f2a055e4d6c998cf8257c87d052b7/statuses?per_page=50\": net/http: invalid header field value \"Bearer \\xf7w\\xdfq\\xaf9m\\xafwkΞ}\\u07b6\\xd5\\xce}\\xe1ל\\xe3]6\\xf3\\x8f\\x1cٷ{\" for key Authorization"}...`

**Comment by simongottschlag** [comment]
Hi,

Please revoke that API key ASAP.

It seems like the formatting of the secret is wrong here and you need to fix that before you can use it with the notification controller.

**Comment by sairamvoggu1007** [comment]
may I know which secret are you referring?

Docs suggesting to encode secrets with base64 as per doc I encoded and create secrets and referring them

**Comment by simongottschlag** [comment]
Remember that even though you've edited the comment, it is possible to view the history of the comment. Make sure it is revoked and you use a new one.

First of all, if you are unable to access the GitHub API from a pod in the cluster with curl, then it won't work with notification-controller. Make sure it is working before you continue troubleshooting it.

Secondly, the format of your token seemed to have weird encoding. Try looking at your secret (decoding it) and see what the output is. If you have stuff like \x in it, then there may be something wrong with how you copied it (maybe zero char whitespace).

**Comment by simongottschlag** [comment]
Are you using the on-prem version of GitHub and trying to communicate with that API?

**Comment by sairamvoggu1007** [comment]
Yes onprem github. we do not want to reach out to api.github.com

**Comment by simongottschlag** [comment]
Ah! I believe (but not sure) the notification-controller doesn't have support for that yet.

Create a feature request for it here: https://github.com/fluxcd/notification-controller

Most likely not something hard, but whoever builds it needs access to GitHub on-prem to validate.

**Comment by simongottschlag** [comment]
After looking at the code, it seems like the following needs to be used for GitHub Enterprise: https://github.com/google/go-github/blob/51f17753badd669f62e5cc9ead041328e2087f4e/github/github.go#L309

**Comment by sairamvoggu1007** [comment]
Sure, I will raise a feature request as mentioned above

**Comment by Nicklason** [comment]
This is not a problem with Github Enterprise, it is a problem with the Kubernetes secret they made. It is because the token is not base64 encoded.

You can use the command below to encode the token.

`echo -n "token" | base64 -w 0`

**Comment by sairamvoggu1007** [comment]
Actually, few days ago already I tried encoding the token and created the secret and referring it. In that case it was not able to fetch the latest commit. I can retry now again and will share the log

**Comment by sairamvoggu1007** [comment]
**the log before encoding the git token**

{"level":"error","ts":"2021-03-16T02:22:48.760Z","logger":"event-server","msg":"failed to send notification","object":"platform/orderstatusfeedservice","kind":"Kustomization","error":"could not list commit statuses: Get \"https://api.github.com/repos/s-voggu/eks/commits/17e02455855f2a055e4d6c998cf8257c87d052b7/statuses?per_page=50\": net/http: invalid header field value \"Bearer for key Authorization"}...

**the log after encoding the git token**

`{"level":"error","ts":"2021-03-16T21:55:53.219Z","logger":"event-server","msg":"failed to send notification","object":"platform/orderstatusfeedservice","kind" │
│ :"Kustomization","error":"could not list commit statuses: context deadline exceeded"} `

I just decoded the token and it looks good

If we notice the log before encoding at least it was trying to get my repo name and latest commit id but after encoding the token not able to even the see latest commit id in the log.

**Comment by simongottschlag** [comment]
Hi!

Now when the token has the correct encoding, we're hitting the issue with GitHub Enterprise not being supported by the notification-controller.

I do recommend creating the feature request for the notification-controller 😊👍

**Comment by phillebaba** [comment]
I can probably answer why this is not working for GitHub enterprise. The reason is most likely because the provider was never built with self hosted Github in mind. If you look at the part that initiates the Github client it ignores the host.
https://github.com/fluxcd/notification-controller/blob/a61680e4c986103d70afa95c74939699e535bb45/internal/notifier/github.go#L54

Instead the client needs to be specifically created for Github enterprise.
https://github.com/google/go-github/blob/51f17753badd669f62e5cc9ead041328e2087f4e/github/github.go#L309

Thanks @simongottschlag for filling me in, it helped getting all the details. I can create a PR which adds support for GH enterprise if you @sairamvoggu1007 can test it for me.
